File: ServerSide/BallPassSchedulePattern/Players_AiTypes/AttentionWeightPlayerDir/AttentionWeightLayer.py
import numpy as np

class AttentionWeight:
    def __init__(self, softmax_layer):
        """
        初期化
        :param softmax_layer: Softmaxの処理を行うクラスのインスタンス
        """
        self.softmax_layer = softmax_layer
        self.hs = None  # 過去状態
        self.h = None   # 現在の入力状態
        self.cache_hs = None  # キャッシュされた過去状態
        self.cache_hr = None  # キャッシュされた現在の状態
        self.softmax_params = None  # Softmaxのパラメータ

    def forward(self, hs_input, h_input):
        """
        フォワード処理
        :param hs_input: 過去状態 (2D配列: T × H)
        :param h_input: 現在の状態 (1D配列: H)
        :return: アテンションの重み (1D配列: T)
        """
        self.hs = np.array(hs_input)
        self.h = np.array(h_input)

        # 現在の状態をコピー
        hr = np.copy(self.h)

        # 過去状態とのスコアを計算
        s = np.dot(self.hs, hr)

        # スコアをSoftmaxで正規化
        a = self.softmax_layer.forward(s)

        # Softmaxのパラメータを保存
        self.softmax_params = self.softmax_layer.get_params()

        # キャッシュに保存
        self.cache_hs = self.hs
        self.cache_hr = hr

        return a

    def backward(self, da):
        """
        バックワード処理
        :param da: アテンションの重みに対する勾配 (1D配列: T)
        :return: 過去状態と現在状態の勾配 (2D配列: [dhs, dh])
        """
        # SoftmaxLayerにパラメータを再設定
        self.softmax_layer.set_params(self.softmax_params)

        # Softmaxの逆伝播
        ds = self.softmax_layer.backward(da)

        # 勾配を初期化
        T, H = self.cache_hs.shape
        dhs = np.zeros_like(self.cache_hs)
        dh = np.zeros_like(self.cache_hr)

        # 勾配を計算
        for i in range(T):
            dhs[i] = ds[i] * self.cache_hr  # dhs[i] = ds[i] * cache_hr
            dh += ds[i] * self.cache_hs[i]  # dh += ds[i] * cache_hs[i]

        return dhs, dh

File: InstanceSide/BallPassSchedulePattern/Players_AiTypes/AttentionWeightPlayerDir/AttentionWeightLayer.cs
using UdonSharp;
using UnityEngine;

public class AttentionWeight : UdonSharpBehaviour
{
    public SoftmaxLayer softmaxLayer; // SoftmaxLayerをインスペクタからアタッチ
    public RinaNumpy rNp; // RinaNumpyをインスペクタからアタッチ

    private float[][] hs; // 入力の過去状態
    private float[] h;    // 現在の入力状態
    private float[][] cacheHs; // 過去状態のキャッシュ
    private float[] cacheHr;   // 現在状態のキャッシュ
    private float[] softmaxParams; // Softmaxのパラメータ

    public float[] Forward(float[][] hsInput, float[] hInput)
    {
        // 入力データを保持
        hs = hsInput;
        h = hInput;

        // ベクトルの長さを取得
        int T = hs.Length;   // 過去状態の数
        int H = hs[0].Length; // 特徴量の次元

        // 現在の状態をコピー
        float[] hr = rNp.Copy_FloatArray(h);

        // 過去状態との類似度を計算（スコア）
        float[] s = new float[T];
        for (int i = 0; i < T; i++)
        {
            s[i] = rNp.DotProduct_FloatArray_FloatArray(hs[i], hr);
        }

        // スコアをSoftmaxで正規化
        float[] a = softmaxLayer.Forward(s);

        // Softmaxのパラメータを取得して保持
        softmaxParams = softmaxLayer.GetParams();

        // キャッシュに保存
        cacheHs = hs;
        cacheHr = hr;

        // 重み（アテンション）を返す
        return a;
    }

    public float[][] Backward(float[] da)
    {
        // キャッシュから長さを取得
        int T = cacheHs.Length;   // 過去状態の数
        int H = cacheHs[0].Length; // 特徴量の次元

        // SoftmaxLayerにパラメータを再設定
        softmaxLayer.SetParams(softmaxParams);

        // Softmaxの逆伝播
        float[] ds = softmaxLayer.Backward(da);

        // 勾配を初期化
        float[][] dhs = new float[T][];
        float[] dh = rNp.ZerosLike_FloatArray(cacheHr);

        // 勾配を計算
        for (int i = 0; i < T; i++)
        {
            dhs[i] = rNp.Multiply_FloatArray_Float(cacheHr, ds[i]); // dhs[i] = ds[i] * cacheHr
            dh = rNp.Add_FloatArray_FloatArray(dh, rNp.Multiply_FloatArray_Float(cacheHs[i], ds[i])); // dh += ds[i] * cacheHs[i]
        }

        // 過去状態と現在状態の勾配を返す
        return new float[][] { rNp.FlattenFloatArray2D(dhs), dh };
    }
}
